{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "# 1. Preparation"
      ],
      "metadata": {
        "id": "8-Q6STjUSQa4"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "Btff6XzuRjNQ",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title Setup\n",
        "#@markdown Run this code cell first. The button with the circle on the left will execute it.\n",
        "import os\n",
        "import subprocess\n",
        "\n",
        "# Install required modules...\n",
        "## R\n",
        "!pip -q install rpy2==3.5.1\n",
        "%load_ext rpy2.ipython\n",
        "##\n",
        "!curl -L -s -O https://github.com/YujiSue/ysngs/releases/download/ja/ysngs-colab.zip\n",
        "!unzip -q ysngs-colab.zip\n",
        "!pip install -q ./ysngs-main\n",
        "\n",
        "# Environment construcion\n",
        "from ysngs import Config, AppManager, WorkFlow, execCmd, common\n",
        "cfg = Config()\n",
        "cfg.setWorkSpace('/content')\n",
        "cfg.setDefault()\n",
        "cfg.makeDirs()\n",
        "apps = AppManager(cfg)"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Folder Structure\n",
        "\n",
        "Running the above will create the following folders:\n",
        "- MyApp : Installation directory for analysis applications\n",
        "- Sample : Upload directory for sequence data\n",
        "- Script : Directory for other analysis scripts\n",
        "- Reference : Storage directory for reference genomes and databases\n",
        "- Preference : Directory for parameter and configuration files"
      ],
      "metadata": {
        "id": "fUM3UsLaS4WY"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 2. Data Upload\n",
        "\n",
        "Create a folder named \"fastq\" in your Google Drive and save your fastq (fq) or compressed (gz) files there."
      ],
      "metadata": {
        "id": "q2TGO5WwSXge"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Mount Google Drive\n",
        "#@markdown This allows Colab to access Google Drive directly. Please follow the instructions that appear after running this cell.\n",
        "from google.colab import drive\n",
        "drive.mount('/content/drive')"
      ],
      "metadata": {
        "cellView": "form",
        "id": "dqVYmR5cHwUK"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Copy from Google Drive\n",
        "#@markdown Running this will start the copy process. This assumes your data is in the \"fastq\" folder on Google Drive. If your data is elsewhere, please change the path below accordingly.\n",
        "\n",
        "import glob\n",
        "import os\n",
        "drive_directory = '/content/drive/MyDrive/fastq' #@param {type:'string'}\n",
        "res = common.execCmd(f'cp {drive_directory}/* /content/Sample', verbose=True)\n",
        "samples = []\n",
        "files = glob.glob('/content/Sample/*')\n",
        "for file in files:\n",
        "  if file.endswith('gz'):\n",
        "    os.system('gunzip ' + os.path.join('/content/Sample/', file))\n",
        "print('Completed.' if res[0] else 'Failed.')"
      ],
      "metadata": {
        "id": "yNEjjpfOfUoq",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 3. Data Quality Check"
      ],
      "metadata": {
        "id": "7wTirzNzUUto"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Install Sequence Data Quality Analysis and Control Softwares\n",
        "apps.install('fastqc')\n",
        "apps.install('fastp')"
      ],
      "metadata": {
        "id": "17CcUQ5ZUY9E",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Sequence Data Quality Analysis (Initial Check)\n",
        "\n",
        "import glob\n",
        "import os\n",
        "import re\n",
        "samples = []\n",
        "paired_seq = True #@param {'type': 'boolean'}\n",
        "files = glob.glob('/content/Sample/*')\n",
        "for file in files:\n",
        "  (dir, fname) = os.path.split(file)\n",
        "  sample = fname[0:fname.rfind('_')]\n",
        "  if sample not in samples:\n",
        "    samples.append(sample)\n",
        "  odir = os.path.join(f'/content/Output/{sample}/{os.path.splitext(fname)[0]}/primary')\n",
        "  os.makedirs(odir, exist_ok=True)\n",
        "  common.execCmd(f'/content/MyApp/FastQC/fastqc -o {odir} {file} &')\n"
      ],
      "metadata": {
        "id": "F6MjKrQSUv1P",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Adapter Cutting and Low-Quality Read Filtering\n",
        "for sample in samples:\n",
        "  files = glob.glob(f'/content/Sample/{sample}*')\n",
        "  fqfiles = []\n",
        "  for file in files:\n",
        "    fqfiles.append(file)\n",
        "  fqfiles.sort()\n",
        "  if len(fqfiles) != 2:\n",
        "    print(f'[Error] File count for \"{sample}\" is not 2.')\n",
        "    continue\n",
        "\n",
        "  os.chdir(f'/content/Output/{sample}')\n",
        "  cmd = f'''/content/MyApp/fastp \\\n",
        "  -i {os.path.splitext(fqfiles[0])[0]}.cut.fq \\\n",
        "  -I {os.path.splitext(fqfiles[1])[0]}.cut.fq \\\n",
        "  -o {os.path.splitext(fqfiles[0])[0]}.filter.fq \\\n",
        "  -O {os.path.splitext(fqfiles[1])[0]}.filter.fq \\\n",
        "  -w 16'''\n",
        "  common.execCmd(cmd, verbose=True)\n",
        "  os.chdir('/content')"
      ],
      "metadata": {
        "id": "JnndJHybzn6D",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Sequence Data Quality Analysis (Post-Filter)\n",
        "for sample in samples:\n",
        "  files = glob.glob(f'/content/Sample/{sample}*.filter.*')\n",
        "  for file in files:\n",
        "    (dir, fname) = os.path.split(file)\n",
        "    odir = f'/content/Output/{sample}/{os.path.splitext(os.path.splitext(fname)[0])[0]}/filtered'\n",
        "    os.makedirs(odir, exist_ok=True)\n",
        "    common.execCmd(f'/content/MyApp/FastQC/fastqc -o {odir} {file} &')\n"
      ],
      "metadata": {
        "id": "SFOLc770xxCg",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Mapping to Reference Genome"
      ],
      "metadata": {
        "id": "OMk8rm2mS4ih"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Install Mapping Software\n",
        "apps.install('samtools')\n",
        "#apps.install('bwa')\n",
        "%cd /content/temp\n",
        "!wget https://jaist.dl.sourceforge.net/project/bio-bwa/bwa-0.7.17.tar.bz2\n",
        "!tar xvf ./bwa-0.7.17.tar.bz2\n",
        "!sed -i 's/const uint8_t rle_auxtab/extern const uint8_t rle_auxtab/g' /content/temp/bwa-0.7.17/rle.h\n",
        "%cd bwa-0.7.17\n",
        "!make -j16\n",
        "!cp bwa /content/MyApp\n",
        "%cd /content\n",
        "!rm -r /content/temp/*\n",
        "apps.install('picard')\n",
        "apps.install('gatk')"
      ],
      "metadata": {
        "cellView": "form",
        "id": "MXIMm5iXT_Gd"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Download Nematode Genome\n",
        "%cd /content/Reference\n",
        "!wget https://downloads.wormbase.org/species/c_elegans/PRJNA13758/sequence/genomic/c_elegans.PRJNA13758.current.genomic.fa.gz\n",
        "!gunzip c_elegans.PRJNA13758.current.genomic.fa.gz"
      ],
      "metadata": {
        "cellView": "form",
        "id": "989HP8hgSzaQ"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ju-UHKdIcwZX",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title Mapping to Reference\n",
        "import os\n",
        "import glob\n",
        "reference_genome = '/content/Reference/c_elegans.PRJNA13758.current.genomic.fa'\n",
        "reference_label = 'worms'\n",
        "\n",
        "os.chdir('/content/Reference')\n",
        "if not os.path.exists(f'{reference_genome}.fai'):\n",
        "  common.execCmd(f'samtools faidx {reference_genome}')\n",
        "if not os.path.exists(f'/content/Reference/{reference_label}.pac'):\n",
        "  common.execCmd(f'/content/MyApp/bwa index -p {reference_label} {reference_genome}', verbose=True)\n",
        "if not os.path.exists(f'{os.path.splitext(reference_genome)[0]}.dict'):\n",
        "  common.execCmd(f'''/content/MyApp/gatk/gatk CreateSequenceDictionary -R {reference_genome} -O {os.path.splitext(os.path.basename(reference_genome))[0]}.dict''', verbose=True)\n",
        "\n",
        "for sample in samples:\n",
        "  os.makedirs(f'/content/Output/{sample}', exist_ok=True)\n",
        "  inputs = glob.glob(f'/content/Sample/{sample}*.filter.*')\n",
        "  inputs.sort()\n",
        "  common.execCmd(f'''/content/MyApp/bwa mem -t 16 -Y -M \\\n",
        "  -R \"@RG\\\\tID:lib_{sample}\\\\tSM:{sample}\\\\tPL:illumina\" \\\n",
        "  {reference_label} {inputs[0]} {inputs[1]} \\\n",
        "  > /content/Output/{sample}/aligned.sam''', verbose=True)\n",
        "os.chdir('/content')"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Create BAM File\n",
        "import glob\n",
        "import os\n",
        "\n",
        "intermediates = []\n",
        "\n",
        "for sample in samples:\n",
        "  common.execCmd(f'''samtools view -@ 16 -b \\\n",
        "  -o /content/Output/{sample}/raw.bam \\\n",
        "  /content/Output/{sample}/aligned.sam''', verbose=True)\n",
        "\n",
        "  common.execCmd(f'''samtools sort -l1 -T /content/temp \\\n",
        "  -@ 16 -O bam -o /content/Output/{sample}/sorted.bam \\\n",
        "  /content/Output/{sample}/raw.bam''', verbose=True)\n",
        "\n",
        "  common.execCmd(f'''java -jar /content/MyApp/picard.jar MarkDuplicates \\\n",
        "  -I /content/Output/{sample}/sorted.bam \\\n",
        "  -O /content/Output/{sample}/{sample}.bam \\\n",
        "  -M /content/Output/{sample}/{sample}.metric.txt''', verbose=True)\n",
        "\n",
        "  common.execCmd(f'samtools index /content/Output/{sample}/{sample}.bam', verbose=True)\n",
        "\n",
        "  intermediates.append(f'/content/Output/{sample}/aligned.sam')\n",
        "  intermediates.append(f'/content/Output/{sample}/raw.bam')\n",
        "  intermediates.append(f'/content/Output/{sample}/sorted.bam')\n",
        "\n",
        "for im in intermediates:\n",
        "  os.system(f'rm {im}')\n",
        "os.system(f'rm /content/Sample/*')"
      ],
      "metadata": {
        "id": "ofJFkppyGVPU",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Variant Calling"
      ],
      "metadata": {
        "id": "upyULP_nmzV0"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Install and Download Required Software and Databases\n",
        "!pip install dash-bio\n",
        "apps.install('sift')\n",
        "# SIFT Database\n",
        "!curl -L -o /content/Reference/WBcel235.83.zip \"https://sift.bii.a-star.edu.sg/sift4g/public//Caenorhabditis_elegans/WBcel235.83.zip\"\n",
        "%cd /content/Reference\n",
        "!unzip /content/Reference/WBcel235.83.zip\n",
        "\n",
        "# SV detection\n",
        "%cd /content/temp/\n",
        "!curl -L -o slib4colab.tar.gz \"https://firebasestorage.googleapis.com/v0/b/publicstorage-3ef6a.appspot.com/o/slib4colab.tar.gz?alt=media&token=b2f15d7c-7c37-4a3d-8613-2e0c0389b086\"\n",
        "!tar -xzf slib4colab.tar.gz\n",
        "!bash ./slib/installOnColab.sh\n",
        "!git clone https://github.com/YujiSue/Sutoku.git\n",
        "!git clone https://github.com/YujiSue/Moirei.git\n",
        "%cd /content/temp/Sutoku\n",
        "!mkdir -p build\n",
        "%cd /content/temp/Sutoku/build\n",
        "!cmake ..\n",
        "!make -j16\n",
        "!cp -n ./sutoku /content/MyApp\n",
        "%cd /content/temp/Moirei\n",
        "!mkdir -p build\n",
        "%cd /content/temp/Moirei/build\n",
        "!cmake ..\n",
        "!make -j16\n",
        "!cp -n ./moirei /content/MyApp\n",
        "%cd /content/temp/Moirei/plugin/mutagen\n",
        "!mkdir -p build\n",
        "%cd /content/temp/Moirei/plugin/mutagen/build\n",
        "!cmake ..\n",
        "!make\n",
        "!mv ./mutagenfilter.plugin /content/MyApp\n",
        "%cd /content/temp/Moirei/plugin/varhist\n",
        "!mkdir -p build\n",
        "%cd /content/temp/Moirei/plugin/varhist/build\n",
        "!cmake ..\n",
        "!make\n",
        "!mv ./varhistio.plugin /content/MyApp\n",
        "\n",
        "#\n",
        "!curl -L -o /content/Reference/nematode.bin \"https://firebasestorage.googleapis.com/v0/b/publicstorage-3ef6a.appspot.com/o/nematode.bin?alt=media&token=b79539da-980d-492d-9cf6-281e2079dac4\"\n",
        "!curl -L -o /content/Reference/nematode.db \"https://firebasestorage.googleapis.com/v0/b/publicstorage-3ef6a.appspot.com/o/nematode.db?alt=media&token=e72cfaf0-601b-4cd4-93a2-99917e988f80\"\n",
        "# Display chart\n",
        "!curl -L -o /content/Reference/nematode_circref.json \"https://firebasestorage.googleapis.com/v0/b/publicstorage-3ef6a.appspot.com/o/nematode_circref.json?alt=media&token=123ef59b-6969-4a8e-a0f9-fd68eda7313f\"\n",
        "%cd /content"
      ],
      "metadata": {
        "id": "KBh6naZvNuj-",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Variant call\n",
        "for sample in samples:\n",
        "  common.execCmd(f'''/content/MyApp/gatk/gatk HaplotypeCaller \\\n",
        "  -R {reference_genome} \\\n",
        "  -I /content/Output/{sample}/{sample}.bam \\\n",
        "  -O /content/Output/{sample}/{sample}.gatk.g.vcf \\\n",
        "  -ERC GVCF -G StandardAnnotation -G AS_StandardAnnotation -G StandardHCAnnotation''', verbose=True)\n",
        "\n",
        "  common.execCmd(f'''/content/MyApp/gatk/gatk GenotypeGVCFs \\\n",
        "  -R {reference_genome} \\\n",
        "  -V /content/Output/{sample}/{sample}.gatk.g.vcf \\\n",
        "  -O /content/Output/{sample}/{sample}.gatk.vcf''', verbose=True)\n"
      ],
      "metadata": {
        "id": "t_6OFIRr6-MT",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Assess Variant Effect\n",
        "for sample in samples:\n",
        "  common.execCmd(f'''java -jar /content/MyApp/SIFT4G_Annotator.jar -c \\\n",
        "  -i /content/Output/{sample}/{sample}.gatk.vcf \\\n",
        "  -d /content/Reference/WBcel235.83 \\\n",
        "  -r /content/Output/{sample} -t''', verbose=True)\n"
      ],
      "metadata": {
        "id": "o3r5ne_Vl9bM",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Structural Variant Detection Parameter Setup\n",
        "!mkdir -p /content/Preference\n",
        "common.execCmd(f'''/content/MyApp/sutoku template --outdir /content/Preference''', verbose=True)\n",
        "\n",
        "import json\n",
        "param = json.load(open('/content/Preference/param.json'))\n",
        "param['variant-param']['homo'] = True\n",
        "param['paired'] = paired_seq\n",
        "param['depth-bin'] = 4\n",
        "param['detect-sv'] = True\n",
        "param['thread'] = 16\n",
        "json.dump(param, open('/content/Preference/ce_svdetect.json', 'w'))\n"
      ],
      "metadata": {
        "id": "njbj5KcEnQEO",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Create Control Data for Structural Variant Detection\n",
        "\n",
        "#@markdown If you have sequenced the parental strain of the mutant sample, this analysis can create control data to filter out background mutations.<br>\n",
        "#@markdown Otherwise, please skip this cell and run the one below to find mutations directly.\n",
        "common.execCmd(f'''/content/MyApp/sutoku summary \\\n",
        "--bam /content/Output/{sample}/{sample}.bam \\\n",
        "--reference /content/Reference/nematode.bin \\\n",
        "--param /content/Preference/ce_svdetect.json''', verbose=True)\n",
        "\n",
        "common.execCmd(f'''mv /content/Output/{sample}/{sample}.bam.bsm /content/Output/control.bsm''')\n",
        "if os.path.exists('/content/Output/control.bsm'):\n",
        "  print(f'Control data : /content/Output/control.bsm')\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "VNKOHpRsoCHt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Structural Variant Detection\n",
        "#@markdown Check this box if you are using control data to filter out background mutations.<br>use_control = False #@param {type: 'boolean'}<br>\n",
        "use_control = False #@param {type: 'boolean'}\n",
        "#@markdown *Note: The detection of a full-length mitochondrial Duplication when not using control data is a false positive due to the circular structure of the genome, which can be interpreted as a tandem duplication.\n",
        "for sample in samples:\n",
        "  common.execCmd(f'''/content/MyApp/sutoku analyze \\\n",
        "  --bam /content/Output/{sample}/{sample}.bam \\\n",
        "  --reference /content/Reference/nematode.bin \\\n",
        "  {\"--control /content/Output/control.bsm\" if use_control else \"\"}\\\n",
        "  --param /content/Preference/ce_svdetect.json \\\n",
        "  {\"--paired\" if paired_seq else \"\"} \\\n",
        "  --depth-bin 4 --detect-sv --thread 16 --oformat tsv''', verbose=True)\n",
        "#@markdown The result file will be output as 'Output' > [Sample Name] > [Sample Name].bam.variants.tsv. The extension is (.tsv), but it can be opened with Excel.<br>\n",
        "#@markdown *If Colab crashes during execution due to RAM failure, stop structural variant detection and run from the code cell below."
      ],
      "metadata": {
        "id": "a5nlXRjduToO",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Narrowing Down Candidate Causal Genes\n",
        "#@markdown - Filtering for selection of the EMS mutation pattern (G-C=>A-T) and homozygous variants.<br>\n",
        "#@markdown - Variants that do not cause amino acid substitutions are deliberately kept, as they can be used to narrow down regions during histogram creation (described below).<br>\n",
        "#@markdown - Histogram data is set to be exported. Creating a histogram of the remaining variants from the above processing will likely show a high frequency of mutations in a limited region of a specific chromosome. Candidate genes are often located within this high-frequency mutation region, so use this for refinement.<br>\n",
        "#@markdown - The result file will be output as 'Output' > [Sample Name] > [Sample Name].filtered.tsv. The extension is (.tsv), but it can be opened with Excel.\n",
        "\n",
        "caller = 'gatk' #@param {type: 'string'}\n",
        "predictor = 'sift' #@param {type: 'string'}\n",
        "ems_mut = True #@param {type: 'boolean'}\n",
        "enu_mut = False #@param {type: 'boolean'}\n",
        "homo_only = True #@param {type: 'boolean'}\n",
        "oformat = 'tsv' #@param {type: 'string'}\n",
        "export_histogram = True #@param {type: 'boolean'}\n",
        "histogram_bin = '500K' #@param {type: 'string'}\n",
        "for sample in samples:\n",
        "  cmd = f'''\n",
        "  /content/MyApp/moirei variantFilter \\\n",
        "  --reference /content/Reference/nematode.bin \\\n",
        "  --param /content/Preference/param.json \\\n",
        "  {\"--homo-only\" if homo_only else \"\"} \\\n",
        "  --plugin-filter /content/MyApp/mutagenfilter.plugin?mutagen={\"EMS\" if ems_mut else (\"ENU\" if enu_mut else \"\")}\\\n",
        "  --plugin-io /content/MyApp/varhistio.plugin?bin={histogram_bin} \\\n",
        "  --oformat {oformat} \\\n",
        "  /content/Output/{sample}/{sample}.gatk_SIFTpredictions.vcf\n",
        "  '''\n",
        "  common.execCmd(cmd, verbose=True)\n",
        "  os.system(f\"mv /content/Output/{sample}/{sample}.gatk_SIFTpredictions.filtered.tsv /content/Output/{sample}/{sample}.filtered.tsv\")\n",
        "  os.system(f\"mv /content/Output/{sample}/{sample}.gatk_SIFTpredictions.hist.json /content/Output/{sample}/{sample}.hist.json\")\n"
      ],
      "metadata": {
        "id": "8sBer-BJKXzz",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Visualization of Variant Distribution\n",
        "import json\n",
        "from dash import Dash, html, dcc, Input, Output, State, callback\n",
        "import dash_bio as dashbio\n",
        "app = Dash(__name__)\n",
        "refdata = json.load(open('/content/Reference/nematode_circref.json'))\n",
        "for sample in samples:\n",
        "  hist = os.path.join('/content/Output', sample, f'{sample}.hist.json')\n",
        "  if not os.path.exists(hist):\n",
        "    continue\n",
        "  data = json.load(open(hist))\n",
        "  for val in data:\n",
        "    val['pos'] = str(val['start']) + '-' + str(val['end'])\n",
        "  layout_config = {\n",
        "      \"labels\": {\n",
        "          \"size\": 24,\n",
        "          \"color\": \"#ffffff\",\n",
        "      },\n",
        "      \"ticks\": {\"display\": False},\n",
        "  }\n",
        "  tracks_config = {\n",
        "      \"innerRadius\": 300,\n",
        "      \"outerRadius\": 400,\n",
        "      \"color\": \"OrRd\",\n",
        "      'tooltipContent': {'name': 'pos'},\n",
        "  }\n",
        "  app.layout = html.Div([\n",
        "      dashbio.Circos(\n",
        "          layout = refdata,\n",
        "          config = layout_config,\n",
        "          tracks = [\n",
        "              {\n",
        "                  \"type\": \"HISTOGRAM\",\n",
        "                  \"data\": data,\n",
        "                  \"config\": tracks_config,\n",
        "              }\n",
        "          ],\n",
        "      )\n",
        "  ])\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "  app.run(debug=True)"
      ],
      "metadata": {
        "id": "n5xQeGdYr48S",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Compress Results and Copy to Google Drive\n",
        "drive_out_directory = '/content/drive/MyDrive/result' #@param {type:'string'}\n",
        "import os\n",
        "os.makedirs(drive_out_directory, exist_ok=True)\n",
        "for sample in samples:\n",
        "  common.execCmd(f\"tar -cvzf {os.path.join('/content/Output', sample + '.tar.gz')} {os.path.join('/content/Output', sample)}\")\n",
        "  res = common.execCmd(f\"cp {os.path.join('/content/Output', sample + '.tar.gz')} {drive_out_directory}\")"
      ],
      "metadata": {
        "cellView": "form",
        "id": "i4XQGxr4GUFC"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}
